src/ht.c

```diff
@@ -1,239 +1,184 @@
-	{ 1048576,	1153459,	1153457   },
-	{ 2097152,	2307163,	2307161   },
-	{ 4194304,	4613893,	4613891   },
-	{ 8388608,	9227641,	9227639   },
-	{ 16777216,	18455029,	18455027  },
-	{ 33554432,	36911011,	36911009  },
-	{ 67108864,	73819861,	73819859  },
-	{ 134217728,	147639589,	147639587 },
-	{ 268435456,	295279081,	295279079 },
-	{ 536870912,	590559793,	590559791 },
-	{ 1073741824,	1181116273,	1181116271},
-	{ 2147483648ul,	2362232233ul,	2362232231ul}
-};
-
-#define entry_is_free(x) (!x || !x->data)
-#define entry_is_deleted(x) (!x || x->data == &deleted_data)
-#define entry_is_present(x) (x && x->data && x->data != &deleted_data)
-
-/**
- * Finds a hash table entry with the given key and hash of that key.
- *
- * Returns NULL if no entry is found.  Note that the data pointer may be
- * modified by the user.
- */
-SdbHashEntry* ht_search(SdbHash *ht, ut32 hash) {
-	ut32 double_hash, hash_address;
-	if (ht && ht->entries) {
-		hash_address = hash % ht->size;
-		do {
-			SdbHashEntry *entry = ht->table + hash_address;
-			if (entry_is_free (entry))
-				return NULL;
-			if (entry_is_present (entry) && entry->hash == hash)
-				return entry;
-			double_hash = hash % ht->rehash;
-			if (double_hash == 0)
-				double_hash = 1;
-			hash_address = (hash_address + double_hash) % ht->size;
-		} while (hash_address != hash % ht->size);
+SdbHash* internal_ht_new(ut64 size, HashFunction hashfunction,
+		ListComparator comparator, DupKey keydup, DupValue valdup,
+		HtKvFreeFunc pair_free) {
+	SdbHash* ht;
+	ut32 i;
+	ht = calloc (1, sizeof(*ht));
+	if (ht != NULL) {
+		ht->size = size;
+		ht->count = 0;
+		ht->load_factor = 1;
+		ht->hashfn = hashfunction;
+		ht->cmp = comparator;
+		ht->dupkey = keydup;
+		ht->dupvalue = valdup;
+		ht->table = calloc (ht->size, sizeof (SdbList*));
+		if (ht->table != NULL) {
+			for (i = 0; i < ht->size; ++i) {
+				ht->table[i] = ls_new ();
+				if (pair_free) {
+					ht->table[i]->free = (SdbListFree)pair_free;
+				}
+			}
+		} else {
+			free (ht);
+			return NULL;
+		}
 	}
-	return NULL;
+	return ht;
 }
 
-static int rehash = 0;
-static void ht_rehash(SdbHash *ht, ut32 new_size_index) {
-	SdbHash old_ht = *ht;
-	SdbHashEntry *e;
-	if (!ht || new_size_index >= ARRAY_SIZE (hash_sizes))
-		return;
-	// XXX: This code is redupped! fuck't
-	ht->table = calloc (hash_sizes[new_size_index].size, sizeof (*ht->table));
-	if (!ht->table)
-		return;
-	rehash = 1;
-	ht->size_index = new_size_index;
-	ht->size = hash_sizes[ht->size_index].size;
-	ht->rehash = hash_sizes[ht->size_index].rehash;
-	ht->max_entries = hash_sizes[ht->size_index].max_entries;
-	ht->entries = 0;
-	ht->deleted_entries = 0;
-	for (e = old_ht.table; e != old_ht.table + old_ht.size; e++) {
-		if (entry_is_present (e))
-			ht_insert (ht, e->hash, e->data, e->iter);
-	}
-	free (old_ht.table);
-rehash = 0;
+SdbHash* ht_new(HashFunction hashfunction, ListComparator comparator,
+		DupKey keydup, DupValue valdup, HtKvFreeFunc pair_free) {
+	ut64 initial_size = 1024;
```

Comment:
why ut64? why 1024? use constants at the begining of the file not in th emiddle of a function